File: studio/consent.py
"""Privacy policy enforcement and consent ledger."""

import hashlib
import json
import os
import time

from studio import config


class PrivacyViolation(Exception):
    pass


_LOCAL_KINDS = frozenset({"ollama", "vllm"})
_CLOUD_KINDS = frozenset({"anthropic", "openai"})


def make_policy(project, run_id):
    """Return providers.get policy callback for this project/run."""
    privacy = project.get("privacy", "local_only")
    consents = set(project.get("stage_consents") or [])

    def policy(stage, kind, model):
        if privacy == "local_only":
            if kind not in _LOCAL_KINDS:
                raise PrivacyViolation(
                    f"local_only: {kind!r} provider blocked for stage {stage!r}"
                )
        elif privacy == "per_stage":
            if kind in _CLOUD_KINDS and stage not in consents:
                raise PrivacyViolation(
                    f"per_stage: stage {stage!r} not in stage_consents"
                )

    return policy


def _append_ledger(project_id, run_id, stage, kind, model, prompt):
    path = config.consent_path(project_id)
    os.makedirs(os.path.dirname(path), exist_ok=True)
    entry = {
        "ts": time.time(),
        "run_id": run_id,
        "stage": stage,
        "kind": kind,
        "model": model,
        "prompt_bytes": len(prompt.encode("utf-8")),
        "prompt_sha256": hashlib.sha256(prompt.encode("utf-8")).hexdigest(),
    }
    with open(path, "a", encoding="utf-8") as f:
        f.write(json.dumps(entry) + "\n")


def wrap_provider(provider, project, run_id, stage, kind, model):
    """Wrap cloud provider generate() to append consent ledger lines."""
    if kind not in _CLOUD_KINDS:
        return provider

    class LedgerWrapper:
        def generate(self, m, prompt, json_mode=False, num_ctx=16384, timeout_s=600):
            _append_ledger(project["id"], run_id, stage, kind, model, prompt)
            return provider.generate(
                m, prompt, json_mode=json_mode, num_ctx=num_ctx, timeout_s=timeout_s
            )

        def healthy(self, m, deadline_s=60):
            return provider.healthy(m, deadline_s=deadline_s)

    return LedgerWrapper()


def read_ledger(project_id):
    path = config.consent_path(project_id)
    if not os.path.exists(path):
        return []
    out = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line:
                out.append(json.loads(line))
    return out

File: src/server/consent.ts
// Port of backend/studio/consent.py — ledger read + runs-engine policy helpers.
import { createHash } from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { consentPath } from "./paths";
import type { Provider, ProviderPolicy } from "./providers";

export class PrivacyViolation extends Error {}

const LOCAL_KINDS = new Set(["ollama", "vllm"]);
const CLOUD_KINDS = new Set(["anthropic", "openai"]);

export type ConsentLedgerEntry = {
  ts: number;
  run_id: string;
  stage: string;
  kind: string;
  model: string;
  prompt_bytes: number;
  prompt_sha256: string;
};

export function readLedger(projectId: string): ConsentLedgerEntry[] {
  const p = consentPath(projectId);
  if (!fs.existsSync(p)) return [];
  const raw = fs.readFileSync(p, "utf-8");
  const out: ConsentLedgerEntry[] = [];
  for (const line of raw.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (trimmed) out.push(JSON.parse(trimmed) as ConsentLedgerEntry);
  }
  return out;
}

export function makePolicy(project: Record<string, unknown>, _runId: string): ProviderPolicy {
  const privacy = String(project.privacy ?? "local_only");
  const consents = new Set((project.stage_consents as string[] | undefined) ?? []);
  return (stage: string, kind: string) => {
    if (privacy === "local_only") {
      if (!LOCAL_KINDS.has(kind)) {
        throw new PrivacyViolation(`local_only: '${kind}' provider blocked for stage '${stage}'`);
      }
    } else if (privacy === "per_stage") {
      if (CLOUD_KINDS.has(kind) && !consents.has(stage)) {
        throw new PrivacyViolation(`per_stage: stage '${stage}' not in stage_consents`);
      }
    }
  };
}

function appendLedger(
  projectId: string,
  runId: string,
  stage: string,
  kind: string,
  model: string,
  prompt: string,
): void {
  const p = consentPath(projectId);
  fs.mkdirSync(path.dirname(p), { recursive: true });
  const entry: ConsentLedgerEntry = {
    ts: Date.now() / 1000,
    run_id: runId,
    stage,
    kind,
    model,
    prompt_bytes: Buffer.byteLength(prompt, "utf-8"),
    prompt_sha256: createHash("sha256").update(prompt, "utf-8").digest("hex"),
  };
  fs.appendFileSync(p, `${JSON.stringify(entry)}\n`, "utf-8");
}

export function wrapProvider(
  provider: Provider,
  project: Record<string, unknown> & { id: string },
  runId: string,
  stage: string,
  kind: string,
  model: string,
): Provider {
  if (!CLOUD_KINDS.has(kind)) return provider;
  return {
    generate(m, prompt, options) {
      appendLedger(project.id, runId, stage, kind, model, prompt);
      return provider.generate(m, prompt, options);
    },
    healthy(m, deadlineS) {
      return provider.healthy(m, deadlineS);
    },
  };
}
